### PR TITLE
add ctest as test driver

### PR DESCRIPTION
**Pull request by felixschurk**
* add `enable_testing()` in top level CMakeLists to discover tests
* add `add_test` for both unit tests as well as integration tests with prefixing the tests with the corresponding test type
* remove custom targets for tests
* remove unneeded warning from https://gitlab.kitware.com/cmake/cmake/-/issues/16062
* adds basic setup for resolving #3445 

Basically one can now invoke the whole test suite with running 
```sh
cmake -S . -B build
cmake --build build --target build_tests
ctest --test-dir build/test 
```
and one does has a quite decent output of the actual test runs.

Parallel running `-j 8`, running specific test cases `-R cpp` or `-R variant`, having minimal progress bar `--progress` as well as repeating tests until failure `--repeat-until-failure 100` is also possible with [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html).

### CODE REVIEW

**Review by ryneeverett — COMMENTED** [review]
By switching from CMD to RUN for the test invocation in the docker images, the tests are now run when the images are *built* and nothing happens when they are *run*.

**Comment by felixschurk** [comment]
Yes that is true, I realized that yesterday but could not fix it on the spot. My knowledge with Docker is limited. 

Would it maybe make sense to move the test invocation out into the `test.yaml` file?
I mean updating this line 
https://github.com/GothenburgBitFactory/taskwarrior/blob/60575a19675713fd5cc209c77733b8ffc3f442e5/.github/workflows/tests.yaml#L105
to 
```yaml
run: docker-compose run test-$CONTAINER ctest --test-dir build/test -j 8 --rerun-failed --output-on-failure
```

This way we would not have duplicated code in all the docker files, however it maybe would be a bit less/or maybe even more clear how we do run the tests.

**Comment by ryneeverett** [comment]
In my opinion reducing the duplication is beneficial and passing the command into the `docker run` invocation is equally clear to having a `CMD` in the image.

**Comment by felixschurk** [comment]
I am a bit sorry for all the failing CI jobs and commit updates, however locally testing it with one image/container it does run. So I am guessing there is a problem with escaping etc. in the `tests.yaml` file.

I had a similar issue already when setting up another CI. (locally working but in GH Actions not).

**Comment by ryneeverett** [comment]
I don't think --entrypoint is what you want. It looks like you were close with 18888160c637e4f9df419475e03d229c4c2351e1. CI failed with "No such service: test-fedora40 ctest". Maybe you just need some quotes so it splits the arguments correctly?
```yml
        run: docker-compose run 'test-${{ env.CONTAINER }}' 'ctest --test-dir build/test -j 8 --output-on-failure --rerun-failed'
```

(Not 100% sure if it should be single or double quotes but I assume the template variables are substituted prior to evaluation and thus don't care.)

**Comment by felixschurk** [comment]
Hm, yes definetely the splitting fixed one problem. 
I then detected some problem with a whitespace (which as somehow different encoded).

Now it does not find some file/directory. Let's see what the problem ist.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,5 @@
 cmake_minimum_required (VERSION 3.22)
 
-# This is a work-around for the following CMake issue:
-# https://gitlab.kitware.com/cmake/cmake/issues/16062
-# The issue has been fixed in CMake 3.11.0; the policy is set
-# to OLD for compatibility with older versions of CMake only.
-if(POLICY CMP0037 AND ${CMAKE_VERSION} VERSION_LESS "3.11.0")
-  cmake_policy(SET CMP0037 OLD)
-endif()
-
```
Comment (djmitche) [inline test/CMakeLists.txt:10]: Thanks for getting rid of this!

**Review by djmitche — COMMENTED** [review]
This is great! I love the more succinct output. A few notes from my experience that don't need to be fixed before this is landed:

 * The `ctest` invocation, and similarly `make -Cbuild test`, don't seem to rebuild tests if their source files have changed, which can lead to surprises when a test run keeps failing after the test has been fixed. Maybe `cmake --build-and-test` helps with this?
 * The tests take a loong time to run. `ctest` takes a `-jN` option just like `make`, which might be a useful thing to recommend (although it did cause `ids.test.py` to fail for me!)
 * Having a documented way to run a single test will be useful, too.

One thing that does need to be addressed here, is updating the docs on how to run the tests (`doc/devel/contrib/development.md`).

**Comment by felixschurk** [comment]
> In my opinion reducing the duplication is beneficial and passing the command into the `docker run` invocation is equally clear to having a `CMD` in the image.

I am sorry, I was not able to work that out. Maybe it makes sense to have this as a new issue (should be an easy one) when someone knows GitHub Actions and Docker.

**Comment by felixschurk** [comment]
Currently, missing is the update of the [test/README](https://github.com/GothenburgBitFactory/taskwarrior/blob/2361521449bc02c9aa0be00dcc98a2d25aadb31b/test/README). 


I also realized now that basically the `--repeat` or respectively `--repeat-until-failure` options from `ctest` do work as a stress test, which we removed in #3447, therefore the functionality is still available, however now a bit more unified.

File: doc/devel/contrib/development.md [hunk 2]
```diff
@@ -49,28 +49,47 @@ cmake --build build-clang
...
-    $ cd build/test
-```
-Then you can run all tests, showing details, with
-```
-    $ make VERBOSE=1 test
-```
-Alternately, run the tests with the details hidden in `all.log`:
+For running the test suite [ctest](https://cmake.org/cmake/help/latest/manual/ctest.1.html) is used.
+Before one can run the test suite the `task_executable` must be built.
+After that also the `build_tests` executable must be build, which can be done over:
```
Comment (djmitche) [inline doc/devel/contrib/development.md]:
```suggestion
After that also the `build_tests` target must be build, which can be done over:
```

File: doc/devel/contrib/development.md [hunk 3]
```diff
@@ -49,28 +49,47 @@ cmake --build build-clang
...
-Either way, you can get a summary of any test failures with:
+This would run all the test in serial and might take some time.
+
+### Running tests in parallel
+```sh
+ctest --test-dir build -j <number-of-jobs>
 ```
-    $ ./problems
+
+Further it is adviced to add the `--output-on-failure` option to `ctest`, to recieve a verbose output if a test is failing as well as the `--rerun-failed` flag, to invoke in subsequent runs only the failed ones.
```
Comment (djmitche) [inline doc/devel/contrib/development.md:71]: Nice!

**Review by djmitche — APPROVED** [review]
Looks good - let's merge it!

**Pull request merged**